src/coreclr/vm/exceptionhandling.cpp

```diff
@@ -4174,6 +4206,179 @@ Exit:;
     return FALSE;
 }
 
+extern "C" CLR_BOOL QCALLTYPE SfiNext(StackFrameIterator* pThis, uint* uExCollideClauseIdx, CLR_BOOL* fUnwoundReversePInvoke, CLR_BOOL* pfIsExceptionIntercepted)
+{
+    QCALL_CONTRACT;
+
+    CLR_BOOL result = FALSE;
+
+    BEGIN_QCALL;
+
+    Thread* pThread = GET_THREAD();
+    Frame* pFrame = pThread->GetFrame();
+    MarkInlinedCallFrameAsEHHelperCall(pFrame);
+
+    result = SfiNextWorker(pThis, uExCollideClauseIdx, fUnwoundReversePInvoke, pfIsExceptionIntercepted);
+
+    END_QCALL;
+
+    return result;
+}
+
+const uint32_t MaxTryRegionIdx = 0xFFFFFFFFu;
+
+static uint32_t CalculateCodeOffset(PCODE pbControlPC, IJitManager::MethodRegionInfo *pMethodRegionInfo)
+{
+    uint codeOffset = (uint)(pbControlPC - pMethodRegionInfo->hotStartAddress);
+    // If the PC is in the cold region, adjust the offset to be relative to the start of the method.
+    if ((pMethodRegionInfo->coldSize != 0) && (codeOffset >= pMethodRegionInfo->hotSize))
+    {
+        codeOffset = (uint32_t)(pMethodRegionInfo->hotSize + (size_t)(pbControlPC - pMethodRegionInfo->coldStartAddress));
+    }
+
+    return codeOffset;
+}
+
+static void InvokeSecondPass(ExInfo *pExInfo, uint idxStart, uint idxLimit)
+{
+    ExtendedEHClauseEnumerator ehEnum;
+    IJitManager::MethodRegionInfo methodRegionInfo;
+
+    if (!EHEnumInitFromStackFrameIterator(&pExInfo->m_frameIter, &methodRegionInfo, &ehEnum))
+        return;
+
+    PCODE pbControlPC = GetControlPC(pExInfo->m_frameIter.m_crawl.GetRegisterSet());
+
+    uint codeOffset = CalculateCodeOffset(pbControlPC, &methodRegionInfo);
+
+    uint lastTryStart = 0, lastTryEnd = 0;
+
+    // Search the clauses for one that contains the current offset.
+    RhEHClause ehClause;
+    for (uint curIdx = 0; EHEnumNextWorker(&ehEnum, &ehClause, /* onlyFinallys */ true) && curIdx < idxLimit; curIdx++)
```

Comment:
I don't believe the `onlyFinallys` logic is actually a good way to implement this. `idxLimit` is based on the index Including the catch clauses, but this will search a different amount. I would instead suggest that the parameter be something like `doNotCalculateCatchType`, and just have the function fill in the `_pTargetType` will NULL if its set to true.